File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Meta/RulesSelfExplain.swift
import Foundation
import Harmonize
import Quick
import XCTest

final class RulesSelfExplain: QuickSpec {
    override class func spec() {
        Given("A Harmonize rule file") {
            let rulesDirectory = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent()
            let rules = swiftFiles(in: rulesDirectory).filter { !infrastructure.contains($0.lastPathComponent) }

            Then("It explains itself with a Rule") {
                let violations = rules.filter { file in
                    let contents = (try? String(contentsOf: file, encoding: .utf8)) ?? ""
                    return !contents.contains("Rule(")
                }
                XCTAssertTrue(violations.isEmpty, "\(rule.message)\n\nViolating files:\n\(violations.map(\.lastPathComponent).joined(separator: "\n"))")
            }
        }
    }

    private static let infrastructure = ["GherkinUtils.swift", "HarmonizeUtils.swift"]

    private static func swiftFiles(in directory: URL) -> [URL] {
        let enumerator = FileManager.default.enumerator(at: directory, includingPropertiesForKeys: nil)
        return enumerator?.compactMap { $0 as? URL }.filter { $0.pathExtension == "swift" } ?? []
    }

    private static let rule = Rule(
        description: "Every lint rule passes a Rule to its assertion.",
        rationale: "A failing rule must explain why it exists and how to fix it, otherwise the next person works around it instead of fixing the code.",
        fixHint: "Replace the raw string with a Rule(description, rationale, fixHint, badExample, goodExample).",
        badExample: "classes.assertTrue(message: \"Use @Factory\") { ... }",
        goodExample: "classes.assertTrue(rule: rule) { ... }"
    )
}

File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/meta/RulesSelfExplain.kt
package com.perrystreet.woof.konsist.meta

import com.lemonappdev.konsist.api.ext.list.withoutName
import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class RulesSelfExplain : BehaviorSpec() {
    init {
        Given("A Konsist rule file") {
            val rules = KonsistUtils.konsistRulesFiles.withoutName("Assertions", "KonsistUtils", "LintRuleMessage")

            Then("It explains itself with a LintRuleMessage") {
                rules.assertTrue(message = Message) { it.hasImport { import -> import.name == "com.perrystreet.woof.konsist.LintRuleMessage" } }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Every lint rule uses LintRuleMessage for its assertion message.",
            why = "A failing rule must explain why it exists and how to fix it, otherwise the next person works around it instead of fixing the code.",
            howToFix = "Replace the raw string with a LintRuleMessage(rule, why, howToFix, badExample, goodExample).",
            badExample = "classes.assertTrue(additionalMessage = \"Use @Factory\") { ... }",
            goodExample = "classes.assertTrue(message = Message) { ... }",
        )
    }
}
